frame/democracy/src/lib.rs

```diff
@@ -1049,8 +1088,107 @@ pub mod pallet {
 					T::Slash::on_unbalanced(T::Currency::slash_reserved(&who, amount).0);
 				}
 			}
-
 			Self::deposit_event(Event::<T>::ProposalCanceled { prop_index });
+			Self::clear_metadata(MetadataOwner::Proposal(prop_index));
+			Ok(())
+		}
+
+		/// Set the metadata to the external proposal.
+		///
+		/// - `origin`: Must be an external origin and correspond to the proposal vote threshold,
+		///   `ExternalOrigin` for `SuperMajorityApprove`, `ExternalDefaultOrigin` for
+		///   `SuperMajorityAgainst` and `ExternalMajorityOrigin` for `SimpleMajority`.
+		/// - `hash`: The preimage hash of an on-chain stored preimage.
+		#[pallet::weight(T::WeightInfo::set_external_metadata())]
+		pub fn set_external_metadata(origin: OriginFor<T>, hash: PreimageHash) -> DispatchResult {
+			let (_, threshold) = <NextExternal<T>>::get().ok_or(Error::<T>::NoProposal)?;
+			match threshold {
+				VoteThreshold::SuperMajorityApprove => {
+					let _ = T::ExternalOrigin::ensure_origin(origin)?;
+				},
+				VoteThreshold::SuperMajorityAgainst => {
+					let _ = T::ExternalDefaultOrigin::ensure_origin(origin)?;
+				},
+				VoteThreshold::SimpleMajority => {
+					let _ = T::ExternalMajorityOrigin::ensure_origin(origin)?;
+				},
+			};
+			ensure!(T::Preimages::len(&hash).is_some(), Error::<T>::BadMetadata);
```

Comment:
Shouldn't we have a clearer error here? Something like `metadata does not exist`, but succinct

Reply:
Also, does it make sense to move this line to before the `<NextExternal<T>::get()` logic?

Reply:
I update the error.
Preimage::let is the heaviest operation here, this is why I am postponing it.